docs/development/abi.md

```diff
@@ -0,0 +1,219 @@
+# Pyodide Platform ABI
+
+## ABIs
+
+### General
+
+Shared libraries must be linked with `-sSIDE_MODULE=1` or
+
+```
+-sSIDE_MODULE=2 -sEXPORTED_FUNCTIONS=<export list>
+```
+
+Note that the name of each symbol in the list needs to be prefixed with an
+underscode. For the smallest result, it is recommended to link with:
+
+```
+-sSIDE_MODULE=2 -sEXPORTED_FUNCTIONS=["_PyInit_MyCModule1", "_PyInit_MyCModule2]
+```
+
+To force all symbols to be exported, link with `-sSIDE_MODULE=1`.
+
+If `-pthread` is used at compile or link time, the resulting libraries will not
+load.
+
+To compile Rust packages, the following flags must be passed to `rustc`:
+
+```
+-C link-arg=-sSIDE_MODULE=2 -C link-arg=-sWASM_BIGINT
+```
+
+When not invoking `rustc` directly, for instance when using cargo, set these
+into the `RUSTFLAGS` environment variable.
+
+Compiling with `-sSIDE_MODULE=1` will not work with Rust because Rust libraries
+contain a `lib.rmeta` file which is not an object file. Rust produces the
+correct list of exported symbols automatically so this should not be a problem
+in practice.
+
+On Rust libc versions older than 0.2.162, it may be necessary to pass
+`-Z link-native-libraries=no` as a `RUSTFLAG`.
+
+Pyodide is statically linked with the following libraries:
+
+- `libegl.js`
+- `libeventloop.js`
+- `libGL`
+- `libhtml5_webgl.js`
+- `libhtml5.js`
+- `liblz4`
+- `libsdl.js`
+- `libwebgl.js`
+- `libwebsocket.js`
+- `bzip2`
+- `zlib`
+
+All of these come from Emscripten ports and the versions of these libraries are
+determined by the version of Emscripten we build with. Any symbols from these
+static libraries may be used by shared libraries.
+
+### pyodide_2024_0
+
+The Emscripten version is 3.1.58. The Python version is 3.12. Python 3.12 must
+be used at build time. All shared libraries must be linked with `-sWASM_BIGINT`.
+
+By default, C++ libraries are built with exceptions disabled, and `throw` is an
+abort. The same is true for `setjmp`/`longjmp`. To enable exceptions and
+`setjmp`/`longjmp`, `-fexceptions` must be passed at compile time and link time.
+
+Specifying an RPATH is not supported and will cause link errors. The dynamic
+loader has been patched so that all dynamic libraries in a wheel named
+`wheel_name-<tag>.whl` will be loaded as if
+`/lib/python3.12/site-packages/wheel_name.libs` is on the `RPATH`, so any
+dynamic library dependencies should be placed in the wheel in a folder called
+`wheel_name.libs`.
+
+### pyodide_2025_0 (under development)
+
+This section reflects the aspirational ABI for `pyodide_2025_0`. This is all
+subject to change without notice.
+
+The Emscripten version is 4.0.6. The Python version is 3.13. Python 3.13 must be
+used at build time.
+
+By default, C++ libraries are built with exceptions disabled, and `throw` is an
+abort. The same is true for `setjmp`/`longjmp`. To enable exceptions and
+`setjmp`/`longjmp`, `-fwasm-exceptions` must be passed at compile time and link time.
```

Comment:
I'm a little unclear how "by default" applies here. What is applying this default? Is there any practical use case for C++ code that *doesn't* support exceptions?

Reply:
emcc is applying this default. There are many complex C++ code bases that don't use exceptions. For example, v8 doesn't use exceptions. But I agree it is more common to use exceptions. We probably will have `pyodide-build` pass `-fwasm-exceptions` by default at some point, I think there is no good reason not to do this. But emcc chooses defaults that make the code small, and then requires you to explicitly opt into functionality that makes the code size bigger.